README.md

```diff
@@ -45,8 +45,25 @@ Use Cases
 ---------
 Two common use cases include
 
-1. [Enforcing programming contracts] throughout a project via runtime checks.
-2. Producing output in `pure` procedures for debugging purposes.
+1. [Producing output in pure procedures] for debugging purposes.
+2. [Enforcing programming contracts] throughout a project via runtime checks.
+
+### Producing output in pure procedures
+Writing pure procedures communicates useful information to a compiler or a developer.
+Specifically, the pure attribute conveys compliance with several constraints that clarify data dependencies and preclude most side effects.
+For a compiler, these constraints support optimizations, including automatic parallelization on a central processing unit (CPU) or offloading to a graphics processing unit (GPU).
+For a developer, the constraints support refactoring tasks such as code movement.
+
+A developer seeking output inside a procedure presumably has an expectation regarding what ranges of output values represent correct program execution.
+A developer can state such expectations in an assertion such as `call_assert(i>0 .and. j<0)`.
+Enforce the assertion by defining the `ASSERTIONS` macro when compiling.
+If the expectation is not met, the program error terminates and prints a stop code showing the assertion's file and line location and a description.
+By default, the description is the literal text of what was asserted: `i>0 .and. j<0` in the aforementioned example.
+Alternatively, the user can provide a custom description.
+
+For richer diagnostic messages from failed assertions, please see the [Julienne] correctness-checking framework.
+Julienne wraps Assert and defines idioms that automatically generate diagnostic messages containing program data.
+Julienne also offers string-handling utilities to assist users with customizing diagnostic messages by, for example, converting an array of numeric type into string representing comma-separated values as text.
```

Comment:
I think we've agreed top move this paragraph further down in this file.

Maybe create a new "Related packages" section at the end ? We could also use such a section to link some of the major consumers of Assert.